mem-cache: Prefetchers Improvements

rjc-arch opened the pull request:
This pull request contains a set of small patches which fix some bugs in the gem5 prefetchers, and aligns out-of-the box prefetcher performance more closely with that which a typical user would expect.

The performance patches have been tested with an out-of-the-box (untuned) Stride prefetcher configuration against a set of SPEC 2017 SimPoints, and show a modest IPC uplift across the board, with no IPC degradation.

The new defaults were identified as part of work on gem5 prefetchers undertaken by Nikolaos Kyparissas while on internship at Arm.

Commit:
mem: added distance parameter to stride prefetcher

The Stride Prefetcher will skip this number of strides ahead of the
first identified prefetch, then generate `degree` prefetches at
`stride` intervals. A value of zero indicates no skip (i.e. start
prefetching from the next identified prefetch address).

This parameter can be used to increase the timeliness of prefetches by
starting to prefetch far enough ahead of the demand stream to cover
the memory system latency.

[Richard Cooper <richard.cooper@arm.com>:
- Added detail to commit comment and `distance` Param documentation.
- Changed `distance` Param from `Param.Int` to `Param.Unsigned`.
]

Change-Id: I6c4e744079b53a7b804d8eab93b0f07b566f0c08
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Signed-off-by: Richard Cooper <richard.cooper@arm.com>

Commit:
mem-cache: Added clean eviction check for prefetchers.

pkt->req->isCacheMaintenance() would not include a check
for clean eviction before notifying the prefetcher,
causing gem5 to crash.

Change-Id: I1d082a87a3908b1ed46c5d632d45d8b09950b382
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

rjc-arch committed:
mem-cache: Update default prefetch options.

Update the default prefetch options to achieve out-of-the box
prefetcher performance closer to that which a typical user would
expect. Configurations that set these parameters explicitly will be
unaffected.

The new defaults were identified as part of work on gem5 prefetchers
undertaken by Nikolaos Kyparissas while on internship at Arm.

Change-Id: Id63868c7c8f00ee15a0b09a6550780a45ae67e55
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

rjc-arch committed:
mem-cache: Squash prefetch queue entries by block address.

Prefetch queue entries were being squashed by comparing the address
of each queued prefetch against the block address of the demand
access. Only prefetches that happen to fall on a cache-line block
boundary would be squashed.

This patch converts the prefetch addresses to block addresses before
comparison.

Change-Id: I55ecb4919e94ad314b91c7795bba257c550b1528
Reviewed-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra committed:
mem-cache: Optimize strided prefetcher address generation

This commit optimizes the address generation logic in the strided
prefetcher by introducing the following changes

(d is the degree of the prefetcher)

* Evaluate the fixed prefetch_stride only once (and not d-times)
* Replace 2d multiplications (d * prefetch_stride and distance *
prefetch_stride) with additions by updating the new base prefetch
address while looping

Change-Id: I49c52333fc4c7071ac3d73443f2ae07bfcd5b8e4
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Richard Cooper <richard.cooper@arm.com>
Reviewed-by: Tiberiu Bucur <tiberiu.bucur@arm.com>

BobbyRBruce reviewed (dismissed):
Though I have little knowledge of the prefetches, I can't find anything wrong here. Nothing here appears capable of breaking anything for current users.

So it's a thumbs-up from me but i'll let it sit a while so others can have a look.

powerjg commented on src/mem/cache/prefetch/Prefetcher.py:
```diff
@@ -192,6 +192,7 @@ class StridePrefetcher(QueuedPrefetcher):
     use_requestor_id = Param.Bool(True, "Use requestor id based history")
 
     degree = Param.Int(4, "Number of prefetches to generate")
+    distance = Param.Int(0, "How far ahead of the demand stream to prefetch")
```
Could you add a sentence here about what "0" means? Does 0 mean the *next* block or the current block?

If "0" means the current block, then we should probably set the default to "1", right?

powerjg reviewed (commented):
Everything looks fine to me. Just one small comment/question below.

rjc-arch commented:
Thanks for the comments @BobbyRBruce & @powerjg! I've updated the comment on the `distance` `Param` as requested, and added some more detail to the commit message.
I've also changed the `Param` to a `Param.Unsigned`, as the distance should be zero or positive.
I've re-tested the updated set of patches on a sample of SimPoints and the changes continue to work as expected.

powerjg reviewed (approved):
Thanks for the updates!

odanrc commented:
A bit late to the party, but LGTM
